service/controller/v3/resource/instance/create.go

```diff
@@ -33,187 +49,205 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 	}
 
-	{
-		d, err := deploymentsClient.Get(ctx, key.ClusterID(customObject), vmssDeploymentName)
-		if IsDeploymentNotFound(err) {
-			// fall through
+	if !resourceStatusExists(customObject, Stage) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring deployment")
+
+		computedDeployment, err := r.newDeployment(ctx, customObject, nil)
+		if controllercontext.IsInvalidContext(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "missing dispatched output values in controller context")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "did not ensure deployment")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			return nil
 		} else if err != nil {
 			return microerror.Mask(err)
 		} else {
-			s := *d.Properties.ProvisioningState
-			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deployment is in state '%s'", s))
-
-			if !key.IsFinalProvisioningState(s) {
-				r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
-				return nil
+			_, err := deploymentsClient.CreateOrUpdate(ctx, key.ClusterID(customObject), vmssDeploymentName, computedDeployment)
+			if err != nil {
+				return microerror.Mask(err)
 			}
-		}
-	}
 
-	// TODO this is a hack we have to fix as soon as the statusresource works and
-	// is enabled.
-	//
-	// Issue: https://github.com/giantswarm/giantswarm/issues/3822
-	//
-	masterVersionsValue := map[string]string{}
-	workerVersionsValue := map[string]string{}
-	{
-		var k8sClient kubernetes.Interface
-		{
-			i := key.ClusterID(customObject)
-			e := key.ClusterAPIEndpoint(customObject)
-			k8sClient, err = r.guestCluster.NewK8sClient(ctx, i, e)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "ensured deployment")
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("setting resource status to '%s/%s'", Stage, DeploymentInitialized))
+
+			err = r.setResourceStatus(customObject, Stage, DeploymentInitialized)
 			if err != nil {
 				return microerror.Mask(err)
 			}
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("set resource status to '%s/%s'", Stage, DeploymentInitialized))
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			return nil
 		}
+	}
 
-		o := metav1.ListOptions{}
-		list, err := k8sClient.CoreV1().Nodes().List(o)
-		if guest.IsAPINotAvailable(err) {
-			// fall through
-			r.logger.LogCtx(ctx, "level", "debug", "message", "not fetching node versions")
-			r.logger.LogCtx(ctx, "level", "debug", "message", "guest API is not available")
+	if hasResourceStatus(customObject, Stage, DeploymentInitialized) {
+		d, err := deploymentsClient.Get(ctx, key.ClusterID(customObject), vmssDeploymentName)
+		if IsDeploymentNotFound(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "deployment not found")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "waiting for creation")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
```

Comment:
Resource is canceled without using `resourcecanceledcontext` is that intended ?

Reply:
This is documented in https://github.com/giantswarm/operatorkit/blob/master/docs/control_flow_primitives.md#cancel-resources. This here is the normal `Resource` implementation. No need for context cancelation. Simple `return` does the trick.